File: python/leetcode/362.py
class HitCounter(object):
    def __init__(self):
        from collections import deque
        self.cnts = 0
        self.hits = deque()

    def hit(self, timestamp):
        if not self.hits or self.hits[-1][0] != timestamp:
            self.hits.append([timestamp, 1])
        else:
            self.hits[-1][1] += 1
        self.cnts += 1

    def getHits(self, timestamp):
        while self.hits and self.hits[0][0] <= timestamp - 300:
            self.cnts -= self.hits.popleft()[1]
        return self.cnts

File: cpp/leetcode/362.cpp
class HitCounter {
private:
    queue<pair<int, int> >  hits; // timestamp:hits
    int cnt; // size 
public:
    HitCounter() {
        cnt = 0;
    }
    
    void hit(int timestamp) {
        // maintain the queue to get rid of the outdated hit pair 
        while(!hits.empty() && timestamp - hits.front().first >= 300 ){
            cnt -= hits.front().second;
            hits.pop();
        }
        // count the current hit:
        ++ cnt;
        if(!hits.empty() && timestamp == hits.back().first)  ++(hits.back().second);
        else    hits.push(make_pair(timestamp,1));
    }
    
    /** Return the number of hits in the past 5 minutes. */
    int getHits(int timestamp) {
        while(!hits.empty() && timestamp - hits.front().first >= 300){
            cnt -= hits.front().second;
            hits.pop();
        }
        return cnt;
    }
};